biom/table.py

```diff
@@ -1599,20 +1599,32 @@ def to_utf8(i):
         else:
             output = ['# Constructed from biom file',
                       '%s%s%s' % (observation_column_name, delim, samp_ids)]
+
+        if direct_io is not None:
+            direct_io.writelines([i+"\n" for i in output])
+
         obs_metadata = self.metadata(axis='observation')
-        for obs_id, obs_values in zip(self.ids(axis='observation'),
+        iterable = self.ids(axis='observation')
+        for obs_id, obs_values in zip(iterable,
                                       self._iter_obs()):
             str_obs_vals = delim.join(map(str, self._to_dense(obs_values)))
-
             obs_id = to_utf8(obs_id)
             if header_key and obs_metadata is not None:
                 md = obs_metadata[self._obs_index[obs_id]]
                 md_out = metadata_formatter(md.get(header_key, None))
-                output.append(
-                    u'%s%s%s\t%s' %
-                    (obs_id, delim, str_obs_vals, md_out))
+                if direct_io is None:
+                    output.append(
+                        u'%s%s%s\t%s' %
```

Comment:
Would it be possible to do this formatting and construction a single time prior to the if statement evaluation?